### PR TITLE
U/jrbogart/polygon

### PR DESCRIPTION
**Pull request by JoanneBogart**
Implement convex polygon region selection.
First excise objects outside bounding box, then use `lsst.sphgeom.ConvexPolygon` for final cut.

### CODE REVIEW

File: python/desc/skycatalogs/skyCatalogs.py [hunk 1]
```diff
@@ -91,26 +185,27 @@ def _compute_mask(region, ra, dec):
...
     '''
     mask = None
-    if type(region) == type(_aBox):
+    if isinstance(region, Box):
         mask = np.logical_or((ra < region.ra_min),
                              (ra > region.ra_max))
         mask = np.logical_or(mask, (dec < region.dec_min))
         mask = np.logical_or(mask, (dec > region.dec_max))
-    if type(region) == type(_aDisk):
+    #if type(region) == type(_aDisk):
```
Comment (jchiang87) [inline python/desc/skycatalogs/skyCatalogs.py:200]: delete this too?

File: python/desc/skycatalogs/skyCatalogs.py [hunk 2]
```diff
@@ -72,8 +132,42 @@ def _compress_via_mask(tbl, id_column, region):
...
+            #t_interval = time.time() - t0
+            #print('Time for trimmed poly mask: ', t_interval)
+            # For any non-zero elts in polygon mask, set corresponding
+            #        elt in full mask array to True.
+            #t0 = time.time()
+            for i, e in enumerate(poly_mask):
+                if e == True:
+                    mask[ixes[i]] = True
+            #t_interval = time.time() - t0
+            #print('Time for merge: ', t_interval)
```
Comment (jchiang87) [inline python/desc/skycatalogs/skyCatalogs.py:167]: I think the following is equivalent and a bit easier to understand (and may be somewhat faster):
```
if all(mask):
    return None, None, None, None

# Compute polygonal region mask for unmasked items.
ra_compress = ma.array(tbl['ra'], mask=mask).compressed()
dec_compress = ma.array(tbl['dec'], mask=mask).compressed()
poly_mask = _compute_mask(region, ra_compress, dec_compress)

# Get indices of unmasked items.
ixes = np.where(~mask)

# Update mask for unmasked items with polygonal region mask.
mask[ixes] = mask[ixes] | poly_mask
```

**Review by jchiang87 — CHANGES_REQUESTED** [review]
One substantive change request, otherwise looks fine.

File: python/desc/skycatalogs/skyCatalogs.py [hunk 3]
```diff
@@ -38,24 +93,29 @@ def _get_intersecting_hps(hp_ordering, nside, region):
...
+    if isinstance(region, Box):
         vec = healpy.pixelfunc.ang2vec([region.ra_min, region.ra_max,
                                         region.ra_max, region.ra_min],
                                        [region.dec_min, region.dec_min,
                                         region.dec_max, region.dec_max],
                                        lonlat=True)
 
         return healpy.query_polygon(nside, vec, inclusive=True, nest=False)
-    if type(region) == type(_aDisk):
+    ###if type(region) == type(_aDisk):
```
[the remark this answers is not in the document]
Comment (JoanneBogart) [reply python/desc/skycatalogs/skyCatalogs.py:104]: yes, of course.

File: python/desc/skycatalogs/skyCatalogs.py (hunk 1 above)
Comment (JoanneBogart) [reply python/desc/skycatalogs/skyCatalogs.py:200]: Likewise.

File: python/desc/skycatalogs/skyCatalogs.py (hunk 2 above)
Comment (JoanneBogart) [reply python/desc/skycatalogs/skyCatalogs.py:167]: Yes, letting numpy do more of the work is a better idea. Thanks.

Comment (JoanneBogart) [reply python/desc/skycatalogs/skyCatalogs.py]: I've implemented this and it seems fine.  I also added some code to the __main__ to try a shape other than a box.   It seems fine. To be sure I made a plot of objects in a diamond inscribed in a box region.
Screen Shot 2022-11-07 at 5 10 41 PM

Comment (JoanneBogart) [reply python/desc/skycatalogs/skyCatalogs.py]: I've rebased to master and added a polygonal region to the row group test.  Ready for review.

File: python/desc/skycatalogs/skyCatalogs.py [hunk 4]
```diff
@@ -75,16 +132,43 @@ def _compress_via_mask(tbl, id_column, region):
     If there are no objects in the region, all return values are None
 
     '''
+    import time
```
Comment (jchiang87) [inline python/desc/skycatalogs/skyCatalogs.py:135]: This line can be deleted.

File: python/desc/skycatalogs/skyCatalogs.py [hunk 5]
```diff
@@ -75,16 +132,43 @@ def _compress_via_mask(tbl, id_column, region):
...
+            ra_min = min([v[0] for v in vertices])
+            dec_max = max([v[1] for v in vertices])
+            dec_min = min([v[1] for v in vertices])
+            bnd_box = Box(ra_min, ra_max, dec_min, dec_max)
+            # Compute mask for that box
+            mask = _compute_mask(bnd_box, tbl['ra'], tbl['dec'])
+            if all(mask): # even bounding box doesn't intersect table rows
+                return None, None, None, None
+            # Save index array (indices for 0's in mask array)
+            ixes = [i for i in range(mask.size) if mask[i] == False]
```
Comment (jchiang87) [inline python/desc/skycatalogs/skyCatalogs.py:150]: These lines can be deleted.

**Review by jchiang87 — CHANGES_REQUESTED** [review]
A couple of trivial changes, otherwise looks good.

File: python/desc/skycatalogs/skyCatalogs.py (hunk 4 above)
Comment (JoanneBogart) [reply python/desc/skycatalogs/skyCatalogs.py]: done

File: python/desc/skycatalogs/skyCatalogs.py (hunk 5 above)
Comment (JoanneBogart) [reply python/desc/skycatalogs/skyCatalogs.py]: done

**Pull request merged**